packages/muelu/src/Misc/MueLu_MultiVectorAggregationTransferFactory_def.hpp

```diff
@@ -0,0 +1,177 @@
+
+    fineLevel.DeclareInput(vectorName, GetFactory("Vector factory").get(), this);
+    Input(fineLevel, "Aggregates");
+    Input(fineLevel, "CoarseMap");
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void MultiVectorAggregationTransferFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level & fineLevel, Level &coarseLevel) const {
+    FactoryMonitor m(*this, "Build", coarseLevel);
+
+    const ParameterList & pL = GetParameterList();
+    std::string vectorName   = pL.get<std::string>("Vector name");
+
+    if (vectorName == "Coordinates")
+      TEUCHOS_TEST_FOR_EXCEPTION(true,Exceptions::RuntimeError,"Use CoordinatesTransferFactory to transfer coordinates instead of MultiVectorAggregationTransferFactory.");
+
+    int numDimensions;
+    RCP<MultiVector> coarseVector;
+    RCP<MultiVector> fineVector;
+
+    if (coarseLevel.IsAvailable(vectorName, this)) {
+      GetOStream(Runtime0) << "Reusing " << vectorName << std::endl;
+      return;
+    }
+
+    RCP<Aggregates> aggregates = Get< RCP<Aggregates> > (fineLevel, "Aggregates");
+    fineVector = fineLevel.Get<RCP<MultiVector> >(vectorName);
+
+    if (aggregates->AggregatesCrossProcessors())
+      TEUCHOS_TEST_FOR_EXCEPTION(true,Exceptions::RuntimeError,"MultiVectorAggregationTransferFactory does not support aggregates that cross processors!");
+
+    // if the coarse map is available, use it. otherwise, construct the map based on aggregates
+    RCP<const Map> coarseMap = Get< RCP<const Map> > (fineLevel, "CoarseMap");
+
+    // coarseMap is being used to set up the domain map of tentative P, and therefore, the row map of Ac
+    // Therefore, if we amalgamate coarseMap, logical nodes in the coordinates vector would correspond to
+    // logical blocks in the matrix
+
+    ArrayView<const GO> elementAList = coarseMap->getLocalElementList();
+
+    LO                  blkSize      = 1;
+    if (rcp_dynamic_cast<const StridedMap>(coarseMap) != Teuchos::null)
+      blkSize = rcp_dynamic_cast<const StridedMap>(coarseMap)->getFixedBlockSize();
+
+    GO                  indexBase    = coarseMap->getIndexBase();
+    size_t              numElements  = elementAList.size() / blkSize;
+    Array<GO>           elementList(numElements);
+
+    // Amalgamate the map
+    for (LO i = 0; i < Teuchos::as<LO>(numElements); i++)
+      elementList[i] = (elementAList[i*blkSize]-indexBase)/blkSize + indexBase;
+
+    RCP<const Map>   uniqueMap      = fineVector->getMap();
+    RCP<const Map>   coarseMVMap = MapFactory        ::Build(coarseMap->lib(), Teuchos::OrdinalTraits<Xpetra::global_size_t>::invalid(), elementList, indexBase, coarseMap->getComm());
+    coarseVector   = Xpetra::MultiVectorFactory<SC,LO,GO,NO>::Build(coarseMVMap, fineVector->getNumVectors());
+
+    // Create overlapped fine vector to reduce global communication
+    RCP<MultiVector> ghostedVector = fineVector;
+    if (aggregates->AggregatesCrossProcessors()) {
```

Comment:
We don't generally support aggregates crossing processors (the aggregation is very expensive), but maybe I don't understand your use case well enough.

Reply:
I agree that we generally don't use/support aggregates that cross processors. This is really just a failsafe so that the algorithm doesn't do something unexpected. For example, in a matrix-free context, I absolutely do not want aggregates to cross processors.